### ISSUE TITLE
Tracking deprected features.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#3986]
|  Feature                                                 | PR(or time)|                                                             | Removed |
|----------------------------------------------------------|------------|-------------------------------------------------------------|----|
| GPU Objectives.                                          | #3643      | Now all objectives having GPU port will use GPU by default. | ✘  |
| `grow_fast_histmaker`                                    | #3836      | Now is renamed to `grow_quantile_histmaker`.                | ✘  |
| Some of R parameters documented in `R/xgb.importance.R`. | #1964      | ??                                                          | ?? |
| `GBTreeModelParam::num_pbuffer_deprecated`               | ??         | ??                                                          | ✘  |
| `python/training.py: train::learning_rates`              | #1797      | Now use use callback API.                                   | ✘  |
| `python/sklearn.py: seed`                                | ??         | Now use random_state.                                       | ✘  |
| `python/sklearn.py: nthread`                             | ??         | Now use n_jobs.                                             | ✘  |
| `XGDMatrixCreateFromCSC`                                 | #1600      | ??                                                          | ✘  |
| `XGDMatrixCreateFromCSR`                                 | #1600      | `XGDMatrixCreateFromCSR_omp`                                | ✘  |
| `DMatrix::SaveToLocalFile`                               | ?? In 2015 | ??                                                          | ✘  |
|                                                          |            |                                                             |    |
|                                                          |            |                                                             |    |

**Comment by trivialfis** [comment]
Inspried by @thvasilo in #3929.(Thanks)  Lets keep track of deprecated features here.  Great chance that I missed something, anyone who notice please help adding it back. :)  @hcho3  Please help filling the `??` if it's still possible (please don't spend too much time on it, it's deprecated), otherwise I will summit a PR to remove everything that loss track.

**Comment by hcho3** [comment]
Let's discuss whether `num_roots` should be kept or not. I think it is not documented and experimental and should be deprecated. Already, I decided not to support it when I first wrote `hist` updater.

**Comment by trivialfis** [comment]
@hcho3 I tried to find out what it does.  It turns out `num_roots` is not documented.  Is there any reference for it?

**Comment by tqchen** [comment]
I think it is fine dropping them, but let us also keep the backward compatibility in mind. It might be useful to first introduce json exchange format, and then allow upgrade via json

**Comment by hcho3** [comment]
Just found out that ``TreeParam.max_depth`` is never set and never used:
https://github.com/dmlc/xgboost/blob/3d81c48d3fa9a2a2ceb0d7388d3ac1436af7d8ee/include/xgboost/tree_model.h#L34

This field is different from the ``TrainParam.max_depth``, which is given at
https://github.com/dmlc/xgboost/blob/42bf90eb8fbcda2fac494c43bc326c7ca1d254d0/src/tree/param.h#L29

Every occurrence of ``max_depth`` refers to ``TrainParam.max_depth``, not ``TreeParam.max_depth``. (Just verified with CLion)

It is not possible to remove ``max_depth`` from ``TreeParam`` right away, since ``TreeParam`` is serialized byte-wise and we cannot change the byte count. However, we can certainly remove ``TreeParam.max_depth`` from the new JSON serialization schema.

**Comment by trivialfis** [comment]
@hcho3 I'm only tracking the features that's **already decided** to be deprecated here,  so that we can have a time frame for removing those code, as you can see some of them are really old and still lingering.

Let's move whether some other features needed to be deprecated into new issues. :)

**Comment by hcho3** [comment]
@trivialfis 

> I'm only tracking the features that's already decided to be deprecated here, so that we can have a time frame for removing those code

Got it. I'll file a new issue to deprecate `TreeParam.max_depth`.

> No idea. The document said so

Turns out `XGDMatrixCreateFromCSC` is indeed deprecated, in favor of `XGDMatrixCreateFromCSCEx`.

**Comment by thvasilo** [comment]
There's an issue on the board about `num_feature` which I guess was used to subsample features? This is taken care of by `colsample*` now so I guess this parameter should be removed. @trivialfis should we open an issue to track this?

**Comment by trivialfis** [comment]
@hcho3 Actually I deprecated a lots of things in #5101, but there is no user visible effect, as all deprecated parameters are duplicated in some way.  I just can't remove them as binary IO is still the default IO method.
